### PR TITLE
Use Spring association resource endpoints for linking entities when a…

### PR DESCRIPTION
**Pull request by aaclan-ebi**
…n input bundle manifest is added to a process

### CODE REVIEW

**Comment by tburdett** [comment]
Hi Alegria, I don't understand this change.  You've switched from using the repository to save to doing a HTTP request to the API from inside the API? This is going to add a lot of network traffic overhead.  It also adds a weird circular dependency that makes it much harder to track exceptions and harder to test. When we add a security to the API this will also be a big maintenance problem.

Is this related to what @rolando-ebi has been talking about? I've told him that I think this is a bad idea before. I also want us to be minimising our use of PATCH requests, not making more, as it increases the likelihood of accidentally stomping over changes that you shouldn't make.  We'll see an increased likelihood of problems like those of state transitions going missing, for example. It's also hard to trace the problems in code.  

To summarise: if this change is to mitigate against concurrent modifications in the Mongo database, the correct way to handle that is not by weakening the semantics of the change (with a PATCH request) and adding an additional circular API dependency, it is by implementing a sensible retry loop (which we will need *anyway*) and locking access to the repository if absolutely necessary (e.g. by using a synchronized block).

**Comment by rolando-ebi** [comment]
@tburdett synchronized won't address the problem here. We would have to make the FileRepository and ProcessesRepository singletons synchronized

**Comment by tburdett** [comment]
So I suggested to synchronize on the file repository.  Why would you need to synchronize on the ProcessesRepository too?

**Comment by rolando-ebi** [comment]
Both attempt .save()s and are at risk of lost updates

**Comment by rolando-ebi** [comment]
Synchronizing on the FileRespository singleton is different to synchzronizing on access to the fileRepository reference in this class. 

To utilise synchronized we would need to synchronize access to FileRepository across the codebase, and make sure that even the generated REST layer has synchronized access to it.

**Comment by tburdett** [comment]
> There is no worry of stomping over a previous request here as the modification is additive
What happens if two analysis pipelines run at the same time? Or if the process is deleted just before the request? Or when we add permissions to file modifications?
> Given that the purpose of this API is to provide a convenient call that replaces multiple REST calls
This isn't really the purpose of this API, the purpose is to encapsulate all the required domain logic for submissions so that clients can take advantage of it, which reduces their complexity.  I don't think the trade-off of implementing that as a secondary call to a lower level API is a good one, for the reasons I gave above
> The network overhead is negligible(the network isn't actually used at all because of the localhost loopback)
At some point we will likely want to exploit horizontal scale out across multiple load balanced instances.  We'll also want to exploit global distribution to support fast submissions too.  What happens then?

**Comment by rolando-ebi** [comment]
> What happens if two analysis pipelines run at the same time? Or if the process is deleted just before the request? Or when we add permissions to file modifications?

We still face all of these issues implementing this using mongo code:
- if the process is deleted just before the request mongo would throw some kind of exception. This implementation would throw a 404 exception.
- when we add permissions to file modifications then this implementation could use the same type of  elevated privileges token used by internal components. Or we could forward the token used in the convenience API request if we wanted that kind of behaviour(we may well want that to ensure consistency)

> This isn't really the purpose of this API

Whether it is or isn't, it is advantageous that we can model this functionality as standard REST calls to the API. This approach minimizes code complexity as it avoids special behaviour i.e direct mongo writes. 

> At some point we will likely want to exploit horizontal scale out across multiple load balanced instances

Mongo replication is what enables horizontal scaling for this functionality. This approach still avails of that.

**Comment by rolando-ebi** [comment]
The only disadvantage of this approach that I can see is that "circular" API argument. I wouldn't call it circular though. It's more akin to a library calling other methods within the library.
